### PR TITLE
Feature: Widgets: Copy-paste, undo-delete

### PR DESCRIPTION
**Pull request by riodeuno**
- [x] Copy selected widget to app storage on mod+c combo press
- [x] Paste copied widget in the next available space at the bottom of the canvas on mod+v combo press
- [x] Allow users to undo a delete widget operation from an action in the toast notification
- [x] Allow users to delete a selected widget using the "del" key
- [x] Allow users to "cut" the selected widget which can be pasted later

### CODE REVIEW

File: app/client/src/actions/widgetActions.tsx [hunk 1]
```diff
@@ -87,3 +87,30 @@ export const forceOpenPropertyPane = (id: string) => {
...
+  };
+};
+
+export const pasteWidget = () => {
+  return {
+    type: ReduxActionTypes.PASTE_COPIED_WIDGET_INIT,
+  };
+};
+
+export const deleteWidget = () => {
```
Comment (hetunandu) [inline app/client/src/actions/widgetActions.tsx:106]: How will this defer from "WidgetDelete" and "WidgetRemoveChild" 
I dont think we should have multiple top level actions to remove a widget

File: app/client/src/components/editorComponents/DropTargetComponent.tsx [hunk 2]
```diff
@@ -240,12 +240,19 @@ export const DropTargetComponent = memo((props: DropTargetComponentProps) => {
...
+      if (!props.parentId) {
+        selectWidget && selectWidget(props.widgetId);
+        focusWidget && focusWidget(props.widgetId);
+      } else {
+        selectWidget && selectWidget(props.parentId);
+        focusWidget && focusWidget(props.parentId);
+      }
       showPropertyPane && showPropertyPane();
     }
+    e.stopPropagation();
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/DropTargetComponent.tsx:254]: Need to test if this blocks any other events propagating through widgets

File: app/client/src/pages/Editor/PropertyPaneTitle.tsx [hunk 3]
```diff
@@ -51,24 +91,85 @@ const PropertyPaneTitle = memo((props: PropertyPaneTitleProps) => {
     setName(props.title);
   }, [props.title]);
 
+  const handleDelete = useCallback(() => dispatch(deleteWidget()), [dispatch]);
```
Comment (hetunandu) [inline app/client/src/pages/Editor/PropertyPaneTitle.tsx:94]: Can we add the Analytics events deleted over here

File: app/client/src/utils/storage.ts [hunk 4]
```diff
@@ -26,3 +28,62 @@ export const hasAuthExpired = async () => {
   }
   return false;
 };
+
+export const saveCopiedWidgets = async (widgetJSON: string) => {
```
Comment (hetunandu) [inline app/client/src/utils/storage.ts:32]: Why do we do is on localstorage and not redux

Comment (riodeuno) [reply app/client/src/utils/storage.ts:32]: This is to allow users to copy from one canvas on one browser tab, and paste it on another canvas in another browser tab. It also, helps us maintain clipboard after a page refresh.

File: app/client/src/components/editorComponents/DropTargetComponent.tsx (hunk 2 above)
Comment (riodeuno) [reply app/client/src/components/editorComponents/DropTargetComponent.tsx:254]: This is only handling the scenario where the user clicks on the canvas (not on any widget). In this case there is no propagation needed, as the click target is user intended. Furthermore, there is no conceptual target to which this click event can propagate.

I have tested this implementation. I will revisit this and test again to confirm.

File: app/client/src/actions/widgetActions.tsx (hunk 1 above)
Comment (riodeuno) [reply app/client/src/actions/widgetActions.tsx:106]: I agree, I'll check if I can re-use or modify other actions which are available.

File: app/client/src/pages/Editor/PropertyPaneTitle.tsx (hunk 3 above)
Comment (riodeuno) [reply app/client/src/pages/Editor/PropertyPaneTitle.tsx]: Analytics events are triggered from the sagas in all widget delete operations.

**Review by hetunandu — APPROVED** [review]

**Pull request merged**